kasa/iot/iottimezone.py

```diff
@@ -22,26 +25,54 @@ async def get_timezone(index: int) -> tzinfo:
     return await CachedZoneInfo.get_cached_zone_info(name)
 
 
-async def get_timezone_index(name: str) -> int:
+async def get_timezone_index(tzone: tzinfo) -> int:
     """Return the iot firmware index for a valid IANA timezone key."""
-    rev = {val: key for key, val in TIMEZONE_INDEX.items()}
-    if name in rev:
-        return rev[name]
+    if isinstance(tzone, ZoneInfo):
+        name = tzone.key
+        rev = {val: key for key, val in TIMEZONE_INDEX.items()}
+        if name in rev:
+            return rev[name]
 
-    # Try to find a supported timezone matching dst true/false
-    zone = await CachedZoneInfo.get_cached_zone_info(name)
-    now = datetime.now()
-    winter = datetime(now.year, 1, 1, 12)
-    summer = datetime(now.year, 7, 1, 12)
     for i in range(110):
-        configured_zone = await get_timezone(i)
-        if zone.utcoffset(winter) == configured_zone.utcoffset(
-            winter
-        ) and zone.utcoffset(summer) == configured_zone.utcoffset(summer):
+        if _is_same_timezone(tzone, await get_timezone(i)):
             return i
     raise ValueError("Device does not support timezone %s", name)
 
 
+async def get_matching_timezones(tzone: tzinfo) -> list[str]:
+    """Return the iot firmware index for a valid IANA timezone key."""
+    matches = []
+    if isinstance(tzone, ZoneInfo):
+        name = tzone.key
+        vals = {val for val in TIMEZONE_INDEX.values()}
+        if name in vals:
+            matches.append(name)
+
+    for i in range(110):
+        fw_tz = await get_timezone(i)
+        if _is_same_timezone(tzone, fw_tz):
+            match_key = cast(ZoneInfo, fw_tz).key
+            if match_key not in matches:
+                matches.append(match_key)
+    return matches
+
+
+def _is_same_timezone(tzone1: tzinfo, tzone2: tzinfo) -> bool:
+    """Return true if the timezones have the same utcffset and dst offset.
+
+    Iot devices only support 109 IANA timezones so this can be used to find
+    other timezones that match the same dst settings by checking they are
+    the same for the current year.
```

Comment:
```suggestion
    As Iot devices only support a static list of IANA timezones, this is used to check if a helper checks if 
    another timezone matches the same dst settings.
    This is needed as XXX
```
Maybe something simpler like this, that describes *why* and *what*? That being said, I'm wondering if we even want to have this sort of magic in the library code, but rather keep the library simple and let the users do as they will?

Reply:
Updated